packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4157,3 +4164,244 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and calls [onChanged] when a
+/// shortcut is triggered that would affect the state of the [controller].
+@visibleForTesting
+class TextEditingHistory extends StatefulWidget {
+  /// Creates an instance of [TextEditingHistory].
+  const TextEditingHistory({
+    Key? key,
+    required this.child,
+    required this.controller,
+    required this.onChanged,
+  }) : super(key: key);
+
+  /// The child widget of [TextEditingHistory].
+  final Widget child;
+
+  /// The [ValueNotifier] to save the state of over time.
+  final TextEditingController controller;
+
+  /// Called when an undo or redo is received, even if the state would be the
+  /// same.
+  final TextEditingValueCallback onChanged;
+
+  @override
+  State<TextEditingHistory> createState() => _TextEditingHistoryState();
+}
+
+// State for [TextEditingHistory].
+class _TextEditingHistoryState extends State<TextEditingHistory> {
+  final UndoStack<TextEditingValue> _stack = UndoStack<TextEditingValue>();
+  late final Throttled<TextEditingValue> _throttledPush;
+  Timer? _throttleTimer;
+
+  // This duration was chosen as a best fit for the behavior of Mac, Linux,
+  // and Windows undo/redo state save durations, but it is not perfect for any
+  // of them.
+  static const Duration _kThrottleDuration = Duration(milliseconds: 500);
+
+  void _undo(UndoTextIntent intent) {
+    _update(_stack.undo());
+  }
+
+  void _redo(RedoTextIntent intent) {
+    _update(_stack.redo());
+  }
+
+  void _update(TextEditingValue? nextValue) {
+    if (nextValue == null) {
+      return;
+    }
+    if (nextValue.text == widget.controller.text) {
+      return;
+    }
+    widget.onChanged(widget.controller.value.copyWith(
+      text: nextValue.text,
+      selection: nextValue.selection,
+    ));
+  }
+
+  void _push() {
+    if (widget.controller.value == TextEditingValue.empty
+        || widget.controller.text == _stack.currentValue?.text) {
+      return;
+    }
+
+    _throttleTimer = _throttledPush(widget.controller.value);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _throttledPush = throttle<TextEditingValue>(
+      duration: _kThrottleDuration,
+      function: _stack.push,
+    );
+    _push();
+    widget.controller.addListener(_push);
+  }
+
+  @override
+  void didUpdateWidget(TextEditingHistory oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
```

Comment:
I am curious whether we may want to reset the history in this case?

If we don't, we may need to call _push() here

Reply:
Good catch.  I'm going to clear the stack here.  I'm not sure if there is a real use case for ever changing the controller like this, but I'm guessing that if there is, the whole field is being reset and the history should reset too.